Big Data And Society
=====


#Introduction to Pandas

##Pandas
Get a Cheatsheet:

from https://drive.google.com/folderview?id=0ByIrJAE4KMTtaGhRcXkxNHhmY2M&usp=sharing


Pandas can read multiple filetypes. CSV files are very common filetypes. 

In [46]:
import pandas as pd
df=pd.read_csv("2013arriyadh_pop_clean.csv")
df.head()

TAZ  Saudi Male  Saudi Female  Total_Saudi  Non-Saudi Male  \
0    1         NaN           NaN          NaN             NaN   
1    2        5956          6943        12899            1943   
2    3        4339          4385         8724            2425   
3    4        1696          1160         2856            2427   
4    5        1710          1553         3263             620   

   Non-Saudi Female  Total_Non_Saudi  Total Pop  
0               NaN              NaN        NaN  
1              2738             4681      17580  
2              1189             3614      12338  
3               106             2533       5389  
4               914             1534       4797

Notice we have a table! A spreadsheet! And it indexed the rows. Pandas (borrowing from R) calls it a DataFrame. Lets see the types of the columns...

`df`, in python parlance, is an instance of the `pd.DataFrame` class, created by calling the `pd.read_csv` function, which calls the DataFrame constructor inside of it. `df` is a dataframe object, and it has methods, or functions belonging to it, which allow it to do things. For example `df.head()` is a method that shows the first 5 rows of the dataframe.

#### DataFrame methods:

In [27]:
df.dtypes

TAZ                 float64
Saudi Male          float64
Saudi Female        float64
Total_Saudi         float64
Non-Saudi Male      float64
Non-Saudi Female    float64
Total_Non_Saudi     float64
Total Pop           float64
dtype: object

The shape of the object is:

In [28]:
df.shape

(1497, 8)

1497 rows times 8 columns. A spredsheet is a table is a matrix. How can we access members of this tuple (brackets like so:() )

In [37]:
df.shape[0], df.shape[1]

(6000, 10)

These are the column names.

In [29]:
df.columns

Index([u'TAZ', u'Saudi Male', u'Saudi Female', u'Total_Saudi', u'Non-Saudi Male', u'Non-Saudi Female', u'Total_Non_Saudi', u'Total Pop'], dtype='object')

###Querying

A spreadsheet is useless if you cant dice/sort/etc it. Here we look for all the TAZ with a Total Populations of less than 300. 

In [30]:
df['Total Pop'] < 300

0      True
1     False
2     False
3     False
4     False
5     False
6      True
7     False
8     False
9     False
10     True
11    False
12    False
13    False
14    False
...
1482     True
1483    False
1484    False
1485    False
1486    False
1487    False
1488     True
1489    False
1490     True
1491     True
1492    False
1493    False
1494    False
1495    False
1496    False
Name: Total Pop, Length: 1497, dtype: bool

This gives us `True`s and `False`s. Such a series is called a mask. If we count the number of `True`s, and divide by the TAZ with a population of less than 300, we'll get the fraction of TAZ $\lt$ 300. To do this numerically see this:

In [33]:
np.sum(df['Total Pop'] < 300)

358

Notice that you could just find the average since the `True`s map to 1s.

In [34]:
np.mean(df['Total Pop'] < 300)

0.23914495657982632

Or directly, in Pandas, which works since `df.rating < 3` is a pandas Series.

In [41]:
(df['Total Pop'] < 300).mean()

0.23914495657982632

###Filtering

Here is a way to get a filtered dataframe: we create a mask and use it to "index" into the dataframe to get the rows we want.

In [42]:
df[df['Total Pop'] < 300]

TAZ  Saudi Male  Saudi Female  Total_Saudi  Non-Saudi Male  \
0        1           0             0            0               0   
6        7          33            29           62              77   
10      11           0             0            0               0   
22      23          35            32           67              19   
25      26           0             0            0               0   
38      39           2             7            9              60   
40      41           3             8           11              81   
47      48          56            45          101              46   
49      50         195            97          292               3   
50      51         109            89          198               2   
59      60           1             1            2               2   
67      68           0             0            0               0   
69      70           2             2            4              11   
74      75           0             0            0               0   
96      97           7             5           12              25   
98      99          25            24           49             107   
110    111           2             1            3               7   
112    113           8             7           15              34   
114    115           3             3            6              42   
116    117           1             0            1               3   
149    150           7             7           14               7   
150    151           0             0            0               0   
160    161           0             0            0               0   
183    184           0             0            0               7   
185    186          10             7           17             132   
214    215          31            34           65             140   
221    222           3             3            6              17   
224    225           0             0            0               0   
225    226          18            12           30              67   
227    228           2             2            4              15   
...    ...         ...           ...          ...             ...   
1366  1367          36            32           68              55   
1396  1397           0             0            0               0   
1397  1398           0             0            0               0   
1401  1402           0             0            0               0   
1404  1405           4             3            7              15   
1405  1406           1             0            1               3   
1411  1412           8             7           15               0   
1412  1413           0             0            0               0   
1413  1414           0             0            0               0   
1416  1417          69            53          122               3   
1431  1432         101            80          181              26   
1434  1435           0             0            0              17   
1436  1437           2             0            2              62   
1440  1441           0             0            0               0   
1441  1442           0             0            0               0   
1442  1443           0             0            0               0   
1443  1444          23            22           45               1   
1444  1445           3             2            5               1   
1445  1446         118           112          230              19   
1447  1448          35            36           71              44   
1453  1454         129           117          246               2   
1455  1456          75            94          169              30   
1456  1457           0             0            0               0   
1465  1466           1             1            2              11   
1466  1467           2             1            3              19   
1472  1473           2             0            2             189   
1482 

If you want to combine conditions, use the second form and put '()' brackets around each condition. The query uses a boolean AND. Each condition ceates a mask of trues and falses.

In [45]:
df[(df['Total Pop'] < 300) & (df['Total_Non_Saudi'] > 70)]

TAZ  Saudi Male  Saudi Female  Total_Saudi  Non-Saudi Male  \
6        7          33            29           62              77   
38      39           2             7            9              60   
40      41           3             8           11              81   
47      48          56            45          101              46   
98      99          25            24           49             107   
185    186          10             7           17             132   
214    215          31            34           65             140   
225    226          18            12           30              67   
267    268           9             8           17              54   
270    271          11            10           21              63   
282    283           7             6           13             173   
306    307          27            22           49             103   
318    319           6             3            9              91   
397    398           5             7           12             129   
410    411           0             0            0              77   
434    435          38            36           74              17   
450    451           0             0            0              83   
451    452           1             0            1             189   
452    453           1             0            1              81   
496    497          35            36           71              49   
509    510           0             0            0             264   
549    550           0             0            0              81   
556    557           0             0            0             175   
564    565           0             0            0             184   
610    611           7             0            7             249   
617    618          45            41           86              53   
784    785          96            85          181              52   
852    853          48            38           86              77   
854    855          25            20           45              88   
931    932          88            77          165              42   
1029  1030           9             7           16              53   
1041  1042          58            57          115              40   
1051  1052          13            14           27              67   
1053  1054          21            22           43             108   
1060  1061           3             2            5              84   
1065  1066           6             4           10              80   
1066  1067          23            18           41              74   
1095  1096          73            73          146              84   
1097  1098          27            19           46              70   
1131  1132          46            49           95              53   
1133  1134          12            13           25              81   
1135  1136           6             6           12             103   
1142  1143          79            48          127              45   
1156  1157          37            25           62             105   
1160  1161          23            15           38              93   
1161  1162          58            39           97              54   
1176  1177          26            18           44             134   
1198  1199          76            69          145              60   
1201  1202          15            14           29             158   
1205  1206          19            18           37             144   
1219  1220           5             4            9             146   
1220  1221           4             3            7             116   
1288  1289          65            43          108             122   
1290  1291          61            40          101             127   
1320  1321          57            49          106             111   
1361  1362          36            32           68             135   
1447  1448          35            36           71              44   
1472 

###Cleaning
We first check the datatypes. 

In [47]:
df.dtypes

TAZ                 float64
Saudi Male          float64
Saudi Female        float64
Total_Saudi         float64
Non-Saudi Male      float64
Non-Saudi Female    float64
Total_Non_Saudi     float64
Total Pop           float64
dtype: object

A common cause of errors is having Null or 'None' values. This usually means data was missing. 

In [51]:
df[df['Total Pop'].isnull()]

TAZ  Saudi Male  Saudi Female  Total_Saudi  Non-Saudi Male  \
0       1         NaN           NaN          NaN             NaN   
1492  NaN         NaN           NaN          NaN             NaN   
1493  NaN         NaN           NaN          NaN             NaN   
1494  NaN         NaN           NaN          NaN             NaN   
1495  NaN         NaN           NaN          NaN             NaN   
1496  NaN         NaN           NaN          NaN             NaN   

      Non-Saudi Female  Total_Non_Saudi  Total Pop  
0                  NaN              NaN        NaN  
1492               NaN              NaN        NaN  
1493               NaN              NaN        NaN  
1494               NaN              NaN        NaN  
1495               NaN              NaN        NaN  
1496               NaN              NaN        NaN

Aha, we had some incomplete data. Lets get rid of it

In [54]:
df = df[df['Total Pop'].notnull()]
df.shape

(1491, 8)

We removed those 6 rows. 

Split-apply-combine
===================

* splitting the data into groups based on some criteria
* applying a function to each group independently
* combining the results into a data structure

Split-apply-combine
===================

<img src=http://i.imgur.com/yjNkiwL.png></img>

Find Diligent Users
===================

* split data per user ID
* count ratings
* combine result

In [99]:
print ratings.head()
## split data
#grouped_data = ratings.groupby('user_id')
grouped_data = ratings['movie_id'].groupby(ratings['user_id'])

## count and combine
#ratings_per_user = grouped_data.count()

#ratings_per_user.head(5)

   user_id  movie_id  rating  unix_timestamp
0      196       242       3       881250949
1      186       302       3       891717742
2       22       377       1       878887116
3      244        51       2       880606923
4      166       346       1       886397596


In [100]:
## split data
grouped_data = ratings['rating'].groupby(ratings['movie_id'])
## average and combine
average_ratings = grouped_data.mean()
print "Average ratings:"
print average_ratings.head()
print

maximum_rating = average_ratings.max()
good_movie_ids = average_ratings[average_ratings == maximum_rating].index

Average ratings:
movie_id
1    3.878319
2    3.206107
3    3.033333
4    3.550239
5    3.302326
Name: rating, dtype: float64



In [13]:
print "Good movie ids:"
print good_movie_ids
print

print "Best movie titles"
print movies[movies.movie_id.isin(good_movie_ids)].title
print

Good movie ids:
Int64Index([814, 1122, 1189, 1201, 1293, 1467, 1500, 1536, 1599, 1653], dtype='int64', name=u'movie_id')

Best movie titles
813                         Great Day in Harlem, A (1994)
1121                       They Made Me a Criminal (1939)
1188                                   Prefontaine (1997)
1200           Marlene Dietrich: Shadow and Light (1996) 
1292                                      Star Kid (1997)
1466                 Saint of Fort Washington, The (1993)
1499                            Santa with Muscles (1996)
1535                                 Aiqing wansui (1994)
1598                        Someone Else's America (1995)
1652    Entertaining Angels: The Dorothy Day Story (1996)
Name: title, dtype: object

